In [4]:
import my_functions as func
import numpy as np
from scipy.stats import f
from scipy.stats import t

In [5]:
normal_distribution_file = open("N.txt", "r")

data = np.array(normal_distribution_file.read().replace('\n', ' ').split(' ')).astype("float64")
X = data[:20]
Y = data[20:]

In [6]:
X_mean = np.mean(X)

In [7]:
X_mean

1.24865

In [8]:
Y_mean = np.mean(Y)

In [9]:
Y_mean

0.8796666666666666

$$ \overline{X} = 1.24865$$
$$ \overline{Y} = 0.87967$$

In [10]:
SX = len(X) * (np.mean(X * X) - np.power(X_mean, 2)) / (len(X) - 1)
SX

0.8913565552631574

$$ S^2_0(X) = 0.8913565552631574 $$

In [11]:
SY = len(Y) * (np.mean(Y * Y) - np.power(Y_mean, 2)) / (len(Y) - 1)
SY

0.8900867816091956

$$ S^2_0(Y) = 0.8900867816091956 $$

$H_1 $= {$\sigma_1^2 = \sigma_2^2$}
$H_2 $= {$\sigma_1^2 \neq \sigma_2^2$}

$$\begin{equation*}
\delta(X, Y) =
 \begin{cases}
   H_1 &,\quad p_{n, m}(X,Y) < q\\
   H_2 &,\quad p_{n, m}(X,Y) \geqslant q
 \end{cases}
\end{equation*}$$
Если верна гипотеза $H_1$
$$ \rho_{20, 30} = \frac{S^2_0(X)}{S^2_0(Y)} \sim F_{19,29}$$
т.е. $$ P(\delta(X, Y) = H1 | H1) = P(\rho_{n, m}<q_1 | H1) = F_{n-1,m-1}(q_1) = 1 - \epsilon$$

Найдем границы для $\epsilon$ = 0.16, и проверим входит ли в них наш $\rho_{20, 30}$

In [12]:
p_20_30 = SX / SY
p_20_30

1.001426572869295

In [13]:
epsilon = 0.16
a = 1
sigma2 = 1.1

In [14]:
q1 = func.find_quantile(1 - epsilon, lambda x: f.cdf(x, X.size - 1, Y.size - 1))
q1

1.4963150024414062

1.001426572869295 < 1.4963150024414062
следовательно гипотеза принимается
Ответ: $ S^2_0(X) = 0.8913565552631574 $, $S^2_0(Y) = 0.8900867816091956 $
$q_1 = 1.4963150024414062,$ $\rho_{20,30}  = 1.001426572869295$


$H_1 $= {$\alpha_1^2 = \alpha_2^2$}
$H_2 $= {$\alpha_1^2 \neq \alpha_2^2$}

$$\begin{equation*}
\delta(X, Y) =
 \begin{cases}
   H_1 &,\quad |p(X,Y)| < q\\
   H_2 &,\quad |p(X,Y)| \geqslant q
 \end{cases}
\end{equation*}$$
Если верна гипотеза $H_1$
$$ \rho(X, Y) = \frac{\overline{X} - \overline{Y}}{ \sqrt{\frac{1}{n}+ \frac{1}{n}} \sqrt{\frac{(n-1)*S^2_X + (m-1)S^2_Y}{n+m-2}} } \sim T_{n+m-2}$$
т.е. $$ P(\delta(X, Y) = H1 | H1) = P(-q<\rho(X, Y)<q | H1) =  T_{n+m-2}(q) - T_{n+m-2}(-q) = 2*T_{n+m-2}(q) - 1  = 1 - \epsilon$$
$$T_{n+m-2}(q) = \frac{2 - \epsilon}{2}$$
Найдем границы для $\epsilon$ = 0.16, и проверим входит ли в них наш $\rho$

In [15]:
n = X.size
m = Y.size
p = (X_mean - Y_mean) / (np.sqrt(1 / n + 1 / m) * np.sqrt(((n - 1) * SX + (m - 1) * SY) / (n + m - 2)))
p

1.3544363849673153

In [16]:
q1 = func.find_quantile((2 - epsilon) / 2, lambda x: t.cdf(x, n + m - 2))
q1

1.42669677734375

$|1.3544363849673153| < |1.42669677734375|$, следовательно гипотеза принимается
Ответ: $q = 1.42669677734375$, $\rho = 1.3544363849673153$